opentelemetry-appender-tracing/src/layer.rs

```diff
@@ -1,60 +1,122 @@
-use opentelemetry::logs::{LogRecord, Logger, LoggerProvider, Severity};
+use opentelemetry::{
+    logs::{AnyValue, LogRecord, Logger, LoggerProvider, Severity},
+    Key,
+};
 use std::borrow::Cow;
-use tracing_core::Level;
+use tracing_core::{Level, Metadata};
+use tracing_log::NormalizeEvent;
 use tracing_subscriber::Layer;
 
 const INSTRUMENTATION_LIBRARY_NAME: &str = "opentelemetry-appender-tracing";
+// Start by "log."
+const LOG_METADATA_ATTRIBUTES: &[&str] = &[
+    "name",
+    "target",
+    "module.path",
+    "file.path",
+    "file.name",
+    "file.line",
+    "module_path",
+    "file",
+    "line",
+];
 
 /// Visitor to record the fields from the event record.
-struct EventVisitor<'a> {
-    log_record: &'a mut LogRecord,
+#[derive(Default)]
+struct EventVisitor {
+    log_record_attributes: Vec<(Key, AnyValue)>,
+    log_record_body: Option<AnyValue>,
 }
 
-impl<'a> tracing::field::Visit for EventVisitor<'a> {
+fn is_metadata(field: &str) -> bool {
+    if let Some(log_field) = field.strip_prefix("log.") {
+        for log_metadata_field in LOG_METADATA_ATTRIBUTES {
+            if *log_metadata_field == log_field {
+                return true;
+            }
+        }
+    }
+    false
+}
+
+fn get_filename(filepath: &str) -> &str {
+    if let Some((_, filename)) = filepath.rsplit_once('/') {
+        return filename;
+    }
+    if let Some((_, filename)) = filepath.rsplit_once('\\') {
+        return filename;
+    }
+    filepath
+}
+
+impl EventVisitor {
+    fn visit_metadata(&mut self, meta: &Metadata) {
+        self.log_record_attributes
+            .push(("log.name".into(), meta.name().into()));
+        self.log_record_attributes
+            .push(("log.target".into(), meta.target().to_owned().into()));
+
+        if let Some(module_path) = meta.module_path() {
+            self.log_record_attributes
+                .push(("log.module.path".into(), module_path.to_owned().into()));
+        }
+        if let Some(filepath) = meta.file() {
+            self.log_record_attributes
+                .push(("log.file.path".into(), filepath.to_owned().into()));
```

Comment:
what does the "log.file.path" indicate here? Is that the line of code where the log was emitted? The [sem. convention doc ](https://opentelemetry.io/docs/specs/semconv/general/logs/#log-file) talks about the file to which log was emitted.

Reply:
Yes it is the source file to which the log was emitted. It is really useful to find a log line in the source code.

Reply:
The line in the source code is in the log.file.line attribute.

Reply:
@Rockin2 From the [documentation ](https://opentelemetry.io/docs/specs/semconv/general/logs/#log-file)it seems attribute `log.file.path` doesn't refer to source code file path, but the destination file path. So if the source file (say) `auth.rs` has code to emit log to `/var/log/auth.log`:
```
log.file.path: /var/log/auth.log
log.file.name: auth.log
```

Reply:
Ouch. You're right. Sorry :(. I will fix it.

Reply:
I think I will rename those log.source.file.path and log.source.file.name.

Reply:
I think the reason I was confused is the log crate to tracing crate bridge add an attribute called "file" that is the path to the source file that emitted the log line.